irr/src/CGLTFMeshFileLoader.cpp

```diff
@@ -0,0 +1,695 @@
+		return std::nullopt;
+	}
+
+	std::vector<video::S3DVertex> vertices;
+	const auto vertexCount = m_gltf_model.accessors->at(*positionAccessorIdx).count;
+	vertices.resize(vertexCount);
+	copyPositions(*positionAccessorIdx, vertices);
+
+	const auto normalAccessorIdx = attributes.normal;
+	if (normalAccessorIdx.has_value()) {
+		copyNormals(normalAccessorIdx.value(), vertices);
+	}
+	// TODO verify that the automatic normal recalculation done in Minetest indeed works correctly
+
+	const auto &texcoords = attributes.texcoord;
+	if (texcoords.has_value()) {
+		const auto tCoordAccessorIdx = texcoords->at(0);
+		copyTCoords(tCoordAccessorIdx, vertices);
+	}
+
+	return vertices;
+}
+
+/**
+ * Get the amount of meshes that a model contains.
+*/
+std::size_t SelfType::MeshExtractor::getMeshCount() const
+{
+	return m_gltf_model.meshes->size();
+}
+
+/**
+ * Get the amount of primitives that a mesh in a model contains.
+*/
+std::size_t SelfType::MeshExtractor::getPrimitiveCount(
+		const std::size_t meshIdx) const
+{
+	return m_gltf_model.meshes->at(meshIdx).primitives.size();
+}
+
+/**
+ * Streams vertex positions raw data into usable buffer via reference.
+ * Buffer: ref Vector<video::S3DVertex>
+*/
+void SelfType::MeshExtractor::copyPositions(
+		const std::size_t accessorIdx,
+		std::vector<video::S3DVertex>& vertices) const
+{
+	const auto accessor = Accessor<core::vector3df>::make(m_gltf_model, accessorIdx);
+	for (std::size_t i = 0; i < accessor.getCount(); i++) {
+		vertices[i].Pos = convertHandedness(accessor.get(i));
+	}
+}
+
+/**
+ * Streams normals raw data into usable buffer via reference.
+ * Buffer: ref Vector<video::S3DVertex>
+*/
+void SelfType::MeshExtractor::copyNormals(
+		const std::size_t accessorIdx,
+		std::vector<video::S3DVertex>& vertices) const
+{
+	const auto accessor = Accessor<core::vector3df>::make(m_gltf_model, accessorIdx);
+	for (std::size_t i = 0; i < accessor.getCount(); ++i) {
+		vertices[i].Normal = convertHandedness(accessor.get(i));
+	}
+}
+
+/**
+ * Streams texture coordinate raw data into usable buffer via reference.
+ * Buffer: ref Vector<video::S3DVertex>
+*/
+void SelfType::MeshExtractor::copyTCoords(
+		const std::size_t accessorIdx,
+		std::vector<video::S3DVertex>& vertices) const
+{
+	const auto accessor = createNormalizedValuesAccessor<2>(m_gltf_model, accessorIdx);
+	const auto count = std::visit([](auto &&a) { return a.getCount(); }, accessor);
+	for (std::size_t i = 0; i < count; ++i) {
+		const auto vals = getNormalizedValues(accessor, i);
+		vertices[i].TCoords = core::vector2df(vals[0], vals[1]);
+	}
+}
+
+/**
+ * This is where the actual model's GLTF file is loaded and parsed by tiniergltf.
+*/
+std::optional<tiniergltf::GlTF> SelfType::tryParseGLTF(io::IReadFile* file)
+{
+	auto size = file->getSize();
+	if (size < 0) // this can happen if `ftell` fails
+		return std::nullopt;
+	auto buf = std::make_unique<char[]>(size + 1);
```

Comment:
Why not `std::string` or `std::vector`?

Reply:
`std::string` is not really intended for use as a mutable buffer (there is no way to get a mutable pointer to the internal data).

`std::vector` would work, but is overkill IMO; we make no use of the capacity here, and we already have the size available. It also requires initialization (well, we could just reserve the right capacity... but still, this is hacky; it's basically just abusing the vector as a weird allocator interface).

But, the way this was written it would zero-initialize as well. Fixed.